### PR TITLE
Issue 1462: (SegmentStore) Reduced number of threads required for container startup

### LINKED ISSUE #1462: Sporadic unit test failure: SegmentStoreIntegrationtests.TimeoutException
**Problem description**
SegmentStoreIntegrationTest.testEndToEnd timed out and SynchronousStreamSegmentStoreTests.testEndToEnd failed

1. io.pravega.segmentstore.server.host.SegmentStoreIntegrationTest > testEndToEnd FAILED
    org.junit.runners.model.TestTimedOutException: test timed out after 300 seconds

2. SynchronousStreamSegmentStoreTests > testEndToEnd failed

**Problem location**
SSS

**Suggestions for an improvement**
bug fix

### PR DESCRIPTION
**Pull request by andreipaduroiu**
**Change log description**
- Introduced `ServiceHelpers`: allows starting and stopping Guava Services and returns a Future that will complete once it's done. This abstracts away all of the async mechanisms built into those services and allows not having to block a thread while waiting for the Service to shut down.
- Making use of `ServiceHelpers` wherever appropriate.
- Retired use of `ServiceShutdownListener` and replaced with `ServiceHelpers`.
- Fixed a bug in `DurableLog.doStart` where it would unnecessarily block an entire thread upon startup.

Other fixes:
- Fixed a bug in `BookKeeperRunner` where it wouldn't set the appropriate ensemble/write/ack quorum size (thus it would be defaulted to 3)

**Purpose of the change**
Fixes #1462. Prior to this, it would require a minimum of 4 available threads to start a Container properly. Now that should only require 1 (albeit it will be slower than having multiple available - but at least it won't hang).

**What the code does**
Every Service startup uses the async listeners to know when they are started or terminated, instead of the sync call previously there.

**How to verify it**
All unit tests must pass.

### CODE REVIEW

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DurableLog.java [hunk 1]
```diff
@@ -147,59 +144,56 @@ protected void doStart() {
...
-                    this.operationProcessor.stopAsync();
-                }
-
-                notifyFailed(ExceptionHelpers.getRealException(ex));
-                return;
-            }
+        // Initiate recovery.
+        CompletableFuture.supplyAsync(this::performRecovery, this.executor)
+                .thenCompose(anyItemsRecovered -> ServiceHelpers.startAsync(this.operationProcessor, this.executor)
+                        .thenComposeAsync(v -> anyItemsRecovered ? CompletableFuture.completedFuture(null) : queueMetadataCheckpoint(), this.executor))
```
Comment (tkaitchuck) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DurableLog.java]: You don't need to nest thenCompose calls you can just chain them.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DurableLog.java [hunk 2]
```diff
@@ -309,7 +303,8 @@ public int getId() {
 
     //region Recovery
 
-    private boolean performRecovery() throws Exception {
+    @SneakyThrows
```
Comment (tkaitchuck) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DurableLog.java]: Looking at the code below, I think wrapping this in a RuntimeException might be more appropriate.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DurableLog.java (hunk 1 above)
Comment (andreipaduroiu) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DurableLog.java]: Ah but I do! You see how I make use of `anyItemsRecovered`? If I chain them I'd lose that one. The alternative would be to store that boolean into an `AtomicBoolean` and reuse it, which would make the code sloppier.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DurableLog.java (hunk 2 above)
Comment (andreipaduroiu) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DurableLog.java]: Hmm, interesting. It looks like both the contents of the `try` and `finally` blocks can throw. I have rewritten the code so it wraps both and rethrows as a `CompletionException`

**Pull request merged**